File: go/0106M-从中序与后序遍历序列构造二叉树.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(inorder []int, postorder []int) *TreeNode {
	if len(postorder) == 0 {
		return nil
	}
	root := &TreeNode{
		Val: postorder[len(postorder)-1],
	}
	if len(postorder) == 1 { // 叶子
		return root
	} 
	var rootIndex int
	for i, v := range inorder { // 中序数组中根的位置
		if v == postorder[len(postorder)-1] {
			rootIndex = i
			break
		}
	}
	// 左闭右开 切割数组，切割后中序、后序数组长度相同
	// inorderLeft := inorder[:rootIndex]
	// inorderRight := inorder[rootIndex+1:]
	// postorderLeft := postorder[:rootIndex]
	// postorderRight := postorder[len(inorderLeft) : len(postorder)-1]

	root.Left = buildTree(inorder[:rootIndex], postorder[:rootIndex])
	root.Right = buildTree(inorder[rootIndex+1:], postorder[rootIndex:len(postorder)-1])
	return root
}

File: cpp/0106M-从中序与后序遍历序列构造二叉树.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-21 14:27:23
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    TreeNode* traversal(vector<int>& inorder, vector<int>& postorder,
                        int inorderBegin, int inorderEnd, int postorderBegin,
                        int postorderEnd) {
        if (postorderBegin >= postorderEnd) return nullptr;
        TreeNode* root = new TreeNode(postorder[postorderEnd - 1]);
        if (postorderEnd - postorderBegin == 1) return root;  // 叶子
        int rootIndex;
        for (int i = inorderBegin; i < inorderEnd; i++) {
            if (inorder[i] == postorder[postorderEnd - 1]) {
                rootIndex = i;
                break;
            }
        }
        root->left = traversal(inorder, postorder, inorderBegin, rootIndex,
                               postorderBegin,
                               rootIndex + postorderBegin - inorderBegin);
        root->right = traversal(inorder, postorder, rootIndex + 1, inorderEnd,
                                rootIndex + postorderBegin - inorderBegin,
                                postorderEnd - 1);
        return root;
    }

    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (postorder.empty() || inorder.empty()) return nullptr;
        return traversal(inorder, postorder, 0, inorder.size(), 0,
                         postorder.size());
    }
};